### DISCUSSION TITLE
Proxy and Implementation ABIs should be imported when creating a new subgraph

### DISCUSSION THREAD

**Discussion by marcusrein** [graphprotocol/graph-tooling#1724] [Ideas]
When indexing a proxy smart contract, automatically find the ABI for the implementation smart contract and spin up the subgraph as appropriately.

One of the big headaches when spinning up a subgraph that is directed at a proxy is to find the ABI of the implementation and manually add it to the subgraph, then manually update the schema, mappings, and subgraph.yaml. Would this be able to be done automatically?

EXAMPLE
```
Ethereum network · iotex-testnet
✔ Contract address · 0x0A7e595C7889dF3652A19aF52C18377bF17e027D
This is a proxy smart contract. Do you want to fetch the ABI from the provided proxy smart contract as well as its implementation smart contract? (Y/n)? True
✔ Fetching proxy ABI from Etherscan
✔ Fetching implementation ABI from Etherscan
✔ Fetching Start Block
✔ Fetching Contract Name
✔ Start Block · 27278816
✔ Contract Name · TransparentUpgradeableProxy
✔ Index contract events as entities (Y/n) · true
  Generate subgraph
  Write subgraph to directory
✔ Create subgraph scaffold
✔ Initialize networks config
✔ Initialize subgraph repository
✔ Install dependencies with yarn
✔ Generate ABI and schema types with yarn codegen
Add another contract? (y/n): n
```

@saihaj and I chatted about this and he added this context:

"I suppose if there are standard proxy contracts interfaces used (like EIP1967) then we could have additional fetching steps.  If there’s not a know format used then it’s hard.
So for the case we detect the proxy l suppose we can get the address to implementation.

There’s some more research that I would need to do in order to figure out proxy impl.
Additionally we would need to solve for our ABI rate limits to make the UX smooth cause today it’s starting to become a real issue.

You should also create an issue and add my above as a context.
I think this one would be a bit more involved effort"

**Comment by YaroShkvorets** [comment]
work in progress

**Reply by marcusrein** [reply]
Thank you!
